File: leetcode/701-insert-into-a-binary-search-tree/InsertIntoABinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        TreeNode nodeToInsert = new TreeNode(val);
        
        // given tree is empty, so node to insert will become a new tree
        if (root == null) {
            return nodeToInsert;
        }
        
        TreeNode prev = null;
        TreeNode node = root;
        while (node != null) {
            prev = node;
            if (node.val > val) {
                node = node.left;
            } else {
                node = node.right;
            }
        }
        
        if (prev.val > val) {
            prev.left = nodeToInsert;
        } else {
            prev.right = nodeToInsert;
        }
        
        return root;
    }
}

File: leetcode/701-insert-into-a-binary-search-tree/insert-into-a-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
var insertIntoBST = function(root, val) {
  if (!root) {
    return new TreeNode(val);
  }

  let ins = false;

  var traverse = function(node) {
    if (ins || !node) {
      return;
    }

    if (!node.left && val < node.val) {
      node.left = new TreeNode(val);
      ins = true;
    } else if (!node.right && val > node.val) {
      node.right = new TreeNode(val);
      ins = true;
    }

    if (val < node.val) {
      traverse(node.left);
    }
    if (val > node.val) {
      traverse(node.right);
    }
  };

  traverse(root);

  return root;
};
